[RNMobile] Image Block: Improve Text Entry for Long Alt Text

SiobhyB opened the pull request:
Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/922

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3236

## Description

With this PR, a new `AltTextSettings` component has been introduced to replace the current `TextControl` that exists for entering an image's Alt text. This new component makes use of `BottomSheet.SubSheet` so that users are taken to a new subsheet when entering alt text.

The end goal is to provide a better experience for users who are entering a longer alt text for their images, as the current text entry has limited room/options for scrolling.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->

The general steps for testing, for both iOS and Android, are as follows:

1. Open **Image Block** settings.
2. Tap **Alt Text** and confirm that a new sub sheet opens with a text editor.
3. Enter text into editor and tap the **back** arrow.
4. Confirm that the image's alt text has been correctly updated by viewing the placeholder displayed alongside **Alt Text** in the settings panel. This can be further verified by saving changes and confirming the alt text has been correctly updated on the web.

As this is a UI change, dark mode and landscape orientation are also considerations when testing.

## Screenshots <!-- if applicable -->

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/110466861-2fe3ab00-80ce-11eb-9d77-a0563cb2c244.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/110466988-5b669580-80ce-11eb-98c1-c7c8c8fe03bf.png" width="100%"> |

## Types of changes

Non-breaking change. No additional functionality is introduced beyond UI enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

SiobhyB committed:
Create new AltTextSettings component

With this commit, a new AltTextSettings component has been introduced to replace the current TextControl that exists for entering an image's Alt text. This new component makes use of BottomSheet.SubSheet, so that users are taken to a new subsheet when entering alt text.

SiobhyB committed:
Update text colour in footer note in styles.css

The current text in the alt text subsheet's footer doesn't meet WordPress' accessibility standards for colour contrast. This commit updates the colour from " $gray-30" to "$gray-text-min" in order to meet those standards.

SiobhyB commented:
Marking as ready for review. :) As well as being keen to hear comments/suggestions on the overall approach I took to address this issue, I'm also curious about the following:

* I used React Native's `TextInput` component instead of one of Gutenberg's components, such as [TextareaControl](https://developer.wordpress.org/block-editor/reference-guides/components/textarea-control/), as it automatically increases in size based on the input. Is that an okay decision to make or is it preferable to use Gutenberg's compontents when possible?
* I made a guess at the colours, based on the `_colors.native.scss` file and also [updated the initial colour](https://github.com/WordPress/gutenberg/pull/29670/commits/a445c7e6df56de890b891174cf1f308a4d70cf81) I chose for the settings to meet WordPress' accessibility standards. Is there a better approach for determining the colours that should be used based on the design?

Thanks in advance!

guarani reviewed (commented):
This is looking solid, functionally it's working great (and I see the `alt` attribute in the editor's HTML). I have some comments to share about the styles (comparing to the design [here](https://github.com/wordpress-mobile/gutenberg-mobile/issues/922#issuecomment-668120593)):

#### Bottom margin too small

When adding alt text, on iOS the margin in-between the text and the horizontal line below looks too tight (some letters almost touch the line). On Android the margin actually looks too big.

| iOS | Android |
| - | - |
| <img width="300" alt="iOS screenshot showing too small text vertical margins" src="https://user-images.githubusercontent.com/1898325/110515226-ea7fa780-80e6-11eb-8b93-839438fb8394.png"> | <img src="https://user-images.githubusercontent.com/1898325/110518791-549a4b80-80eb-11eb-9daa-e80d6ecbf265.jpg" alt="Android screenshot showing too big text vertical margins" width="300"> |

#### Horizontal line and color 

On both iOS and Android, the horizontal line color looks a bit off (especially in dark mode) when comparing to other places in the app (like Cover block bottom sheet):

| Image alt text on iOS | Cover block on iOS |
| -- | -- |
| <img width="350" alt="Screen Shot 2021-03-09 at 15 05 16" src="https://user-images.githubusercontent.com/1898325/110517133-37647d80-80e9-11eb-9424-48060ffbf82d.png"> | <img width="350" alt="Screen Shot 2021-03-09 at 15 05 33" src="https://user-images.githubusercontent.com/1898325/110517152-3df2f500-80e9-11eb-9afa-32e79f7a8345.png"> |

#### Text size and color

Looking at the example design [here](https://github.com/wordpress-mobile/gutenberg-mobile/issues/922#issuecomment-668120593), the text side and color for the "Describe the.." string seems a bit different here. I'm not sure if @iamthomasbishop specified a particular text size that he'd like, but we could probably reduce it a little.

#### Dark mode

On iOS and Android the text stays black on the dark background in my tests:

<img width="350" alt="Screen Shot 2021-03-09 at 14 44 11" src="https://user-images.githubusercontent.com/1898325/110516528-76460380-80e8-11eb-9bb9-e2d83becb062.png">

guarani commented:
>     * I used React Native's `TextInput` component instead of one of Gutenberg's components, such as [TextareaControl](https://developer.wordpress.org/block-editor/reference-guides/components/textarea-control/), as it automatically increases in size based on the input. Is that an okay decision to make or is it preferable to use Gutenberg's compontents when possible?

Just regarding this, it looks like the `TextareaControl` has a native implementation that uses `TextInput`: https://github.com/WordPress/gutenberg/blame/trunk/packages/components/src/textarea-control/index.native.js#L14
Did you run into issues with it not automatically increasing in size when text was multi-line?

iamthomasbishop commented:
Nice work so far, @SiobhyB! 👏  Happy to see this in the works! 

I think @guarani  can lend a better hand with implementing/matching colors, but I'll try to clarify some things regarding sizing/spacing:

> I'm not sure if @iamthomasbishop specified a particular text size that he'd like, but we could probably reduce it a little.

That's right, I believe I used the section footer component in Figma, which we're using various places (i.e. Columns block settings). I can't recall if we have the exact thing as a structured component yet, but it looks like this:

<a href="https://user-images.githubusercontent.com/1200432/110551337-ec536600-80fa-11eb-8ccd-e964155deca5.png"><img src="https://user-images.githubusercontent.com/1200432/110551337-ec536600-80fa-11eb-8ccd-e964155deca5.png" width="798"></a>

Let me clarify what I was imagining for padding/sizing of the text area. If we use the main cell component as a staring point, and expand the text label area downward, we would have something like this:

<a href="https://user-images.githubusercontent.com/1200432/110552718-6edd2500-80fd-11eb-8287-ec745413a72c.png"><img src="https://user-images.githubusercontent.com/1200432/110552718-6edd2500-80fd-11eb-8287-ec745413a72c.png" width="814"></a>

I'm not sure if the `padding-top` is 100% accurate to the production component, but let's use the same value. What's different here is that we are setting a `min-height` for the input — I'm thinking 3 or 4 lines-worth of text, so in the ballpark of ~80px (for the inner part of the block with padding this would == ~108px).

SiobhyB committed:
Increase min-height of altTextEditor

This commit increases the min-height of "altTextEditor" so that the editor is a larger height on iOS devices by default.

SiobhyB committed:
Update border colours for both light and dark mode

This commit updates the component's border colours to match similar separators/borders throughout the app. "light-gray-400" is used for light mode and "gray-70" for dark mode.

SiobhyB committed:
Update text colour in dark mode

This commit fixes a bug, where text entered in iOS appear black on black (therefore, unreadable) in dark mode.

SiobhyB committed:
Update footer note styling in styles.css

With this commit, I've tweaked the font size and padding around the "altTextSetting" compont's footer note to more closely match up with the initial design.

SiobhyB committed:
Set max height for the editor

This commit fixes a bug I spotted while testing, where it's possible to continue hitting "enter" and increasing the size of the editor without limitless, leading to the footer note being dragged off screen.

SiobhyB committed:
Increase bottom padding of footer note

This commit increase the padding at the bottom of the footer note, in accordance with the design.

SiobhyB committed:
Update colour of placeholder text

This commit updates the colour of the default placeholder text in the "AltTextSettings" component's TextInput. The purpose of defining this is to fix a bug in iOS, where the placeholder text appeared as black on black in dark mode.

SiobhyB committed:
Styling tweaks to footnote and placeholder text

Two small changes to the "Alt Text Settings" components are included in this commit. The first increases the padding of the footnote to match the design and the second aligns the placeholder text to the top, which is necessary for Android devices.

SiobhyB commented:
@guarani, @iamthomasbishop, thanks so much for the help here. 🙇‍♀️ I've made updates to change the styles, based on your feedback, and have taken screenshots of the results below.
  
# iOS

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/110645798-32b5cd00-81ae-11eb-85a0-e35a3ae2fab8.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/110645931-537e2280-81ae-11eb-9984-6deda26be326.png" width="100%"> |

# Android

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/110649364-6e9e6180-81b1-11eb-83db-f382dc00cde4.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/110649468-883fa900-81b1-11eb-87ea-319b2762c76f.png" width="100%"> |

# Notes

I found that [the `FooterMessageControl` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/footer-message-control/index.native.js) was used to add a footer to the Columns block settings. I didn't use it for the footer note here as the design was slightly different and I couldn't find a simple way of incorporating a hyperlink using it. So, I instead relied on `Text` components for this purpose. Let me know if that seems okay or if you think another approach is warranted. :) Looking forward to hearing any further feedback, too!

guarani reviewed (approved):
> I made a guess at the colours, based on the _colors.native.scss file and also updated the initial colour I chose for the settings to meet WordPress' accessibility standards. Is there a better approach for determining the colours that should be used based on the design?

I think your approach here is what I've used. Just want to mention that depending on what kind of changes are included in a PR, an installable build can be made for WPiOS and WPAndroid. If this is something that you'd like @iamthomasbishop, I'm available to help show you @SiobhyB how to make those.

Regarding the question of whether we should be using `TextareaControl` instead of a plain `TextInput`, I have no strong feelings either way. The code looks good and I've tested both platforms so this is about done. What's left from my perspective is to add a note to the change log here: https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/CHANGELOG.md

iamthomasbishop commented on packages/components/src/mobile/alt-text-settings/styles.scss:
```diff
@@ -0,0 +1,29 @@
+.altTextEditor {
+	padding: 14px 0;
+	min-height: 80px;
+	max-height: 250px;
```
@SiobhyB Is the text area here scrollable upon overflow? If so, I wonder if it'd be simpler to just set a static `height` (perhaps using that 80px value?) to remove some complexity that might come with expanding/shrinking the input?

SiobhyB replied:
@iamthomasbishop, yes, that's right, the input box currently expands in height when more than four lines of text are added. I've now updated this to a set height of 80px, based on your feedback, so that it no longer expands. :) It's still scrollable when more lines are added.

iamthomasbishop replied:
Sounds good! Hopefully that works nicely 😄

iamthomasbishop commented on packages/components/src/mobile/alt-text-settings/styles.scss:
```diff
@@ -0,0 +1,29 @@
+}
+
+.altTextFootnote {
+	font-size: $default-font-size;
```
From what I can tell, this is a web-defined variable, which I'm not sure is what we'd like here. I'm not if we've got the it implemented quite yet, but in Figma I've typically been using a "Caption" from this [neutral type-scale](https://zeroheight.com/226850145/p/197239-typography/b/24b504), which is 12px. 

Are we using this same variable on the other instances of the section footer component (columns, etc)? If so, let's just use what you have here (`$default-font-size`) and worry about the neutral type-scale at a later time.

SiobhyB replied:
> Are we using this same variable on the other instances of the section footer component (columns, etc)? If so, let's just use what you have here ($default-font-size) and worry about the neutral type-scale at a later time.

[The `FooterMessageControl` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/footer-message-control/index.native.js) being used for the columns block uses a font-size of 15px (with the `$text-editor-font-size` variable) and the text is centred:

![Screenshot_1615444170](https://user-images.githubusercontent.com/2998162/110745588-62f37f00-8233-11eb-89a9-21064d15f999.png)

I found the `$default-font-size` variable in the same file where `$text-editor-font-size` ([`_variables.scss`](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_variables.scss)) is declared, which is how I ended up using it.

I've changed this to 12px now. :)

iamthomasbishop replied:
Thank you! One thing I noticed relative to the screenshots in this other comment — it looks like the text color (not the link) in your PR is a tad darker than in the screenshot above (Columns settings). Can we align our footnote color here with the one on Columns?

SiobhyB replied:
@iamthomasbishop, it looks like the colour contrast in the Columns setting is a bit below what's outlined in the accessibility standards for WordPress. I've gone ahead to open a separate issue for that in https://github.com/WordPress/gutenberg/issues/29866. I ended up using the `$gray-text-min` variable I found in [the _colors.native.scss file](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_colors.native.scss#L50) to get the lightest colour that passes the standards here. Do you think that seems fair or is there perhaps another colour you'd prefer to use? I can submit a PR to also update the colour of the Footer Message component depending on what you think.

iamthomasbishop commented on packages/components/src/mobile/alt-text-settings/styles.scss:
```diff
@@ -0,0 +1,29 @@
+}
+
+.altTextLink {
+	color: $blue-wordpress;
```
I believe this is an old blue (which is not very clear, based on a quick glance at the code 😄 ). For reference, we typically use `blue-50` in light mode and `blue-30` in dark mode for the primary blue.

SiobhyB replied:
Updated and noted for the future, thank you! Here are the updated screenshots of what this looks like in the Android app, including the slightly smaller font-size we now have: 

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/110747916-eebada80-8236-11eb-82d7-dc14d898f41b.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/110748048-1f9b0f80-8237-11eb-84e4-f70fd0449a8a.png" width="100%"> |

iamthomasbishop replied:
Ah, that looks better! Thank you!

iamthomasbishop commented on packages/components/src/mobile/alt-text-settings/styles.scss:
```diff
@@ -0,0 +1,29 @@
+.horizontalBorder {
+	border-top-width: $border-width;
+	border-bottom-width: $border-width;
+	border-color: $light-gray-400;
```
I don't think these border colors are the right ones, but it doesn't look like we have the ones in there that I'd have assumed we could use. 🤔 cc @Guarani Did we ever finish moving all of the border-colors over to the "new" neutral colors? I'm not seeing them in `colors.native.scss` but I was almost positive that we added some color variables for border/separator during last hack week. Maybe they're in a different file?

SiobhyB replied:
For reference, I got these colours from looking at the following other instances of separators within the app:

* https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/bottom-sheet/styles.native.scss#L19
* https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/bottom-sheet/styles.native.scss#L89
* https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/color-settings/style.native.scss#L2
* https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/picker/styles.native.scss#L1

Let me know if there's another way to gather the correct ones!

guarani replied:
We didn't finish @iamthomasbishop 😞 I remember still having some confusion about the changes we were supposed to make. There's a PR open [here](https://github.com/WordPress/gutenberg/pull/25429) and I think the internal discussion was still on-going when we paused this after that HACK week finished.

guarani replied:
The PR I think you're remembering is [this one](https://github.com/WordPress/gutenberg/pull/25350/files). I do think we need regroup and continue this. I remember some confusion about how to map the old colors to the new, so maybe we could start again by creating an issue that specifies the work needed to be done and we go from there, wdyt?
Sorry to discuss a separate issue on this PR.

iamthomasbishop replied:
> For reference, I got these colours from looking at the following other instances of separators within the app:

Ahh, fair enough! I knew we hadn't updated the grays for the separators yet, but I wasn't sure which color was being used. I'd like to go through at some point and make sure all of our separators/borders are using the same "new" color (which as @Guarani pointed out looks like it didn't quite get shipped) but let's leave that for another time. In other words, this is fine for now because it's consistent w/ the other border colors 😄

iamthomasbishop reviewed (changes requested):
@SiobhyB Looking pretty close, for sure! I dug in a bit more and have a few questions/bits of feedback, which I'll try to find/drop inline.

SiobhyB committed:
Update font-size of footnote

Changes the font-size of the footnote from 13px to 12px, based on feedback.

SiobhyB committed:
Set defined height for altTextEditor

This commit sets a defined height of 80px for the altTextEditor component, in response to feedback and to remove complexity around the box expanding when text is entered.

SiobhyB committed:
Update blue colour of link for both light and dark mode

The colour of the link is updated for both light and dark mode with this commit, based on feedback received.

SiobhyB committed:
Update CHANGELOG.md

Update CHANGELOG.md with changes from this commit.

enejb commented:
Nice work @SiobhyB ! 

I haven't had a chance to test the PR yet. But reading the code 
One thing that I am wondering though is should we make the control a bit more generic now or later? 

Do we see this new AltTextSettings control being used by other blocks that face a similar issue? 

If we renamed it to something like `BottomSheetTextControll` and we could pass the different content that is specific to 
current `AltTextSettings` into it would make it easier down the road the implement a similar feature on different blocks in the code base. 

I am not sure if this is the right time to do this :) but something to keep in mind. 

On a related note it would be great to add a quick Readme to the new `AltTextSettings` component this helps other developers know how to use it :)

enejb commented on packages/components/src/mobile/alt-text-settings/styles.scss:
```diff
@@ -0,0 +1,32 @@
+}
+
+.altTextFootnote {
+	font-size: 12px;
```
Is there a sass variable that we can use for this?

SiobhyB replied:
This was changed to 12px following [this discussion](https://github.com/WordPress/gutenberg/pull/29670/files/ff711fa39da42c484e5f0725f284851a28ab14b8#r591931227) and I wasn't able to find a variable for 12px under [the _variables.scss file](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_variables.scss). I was a bit wary of creating a new variable as I thought that might be part of a larger conversation, as I think @iamthomasbishop may have alluded to in one of his replies ("... and worry about the neutral type-scale at a later time"). I'd be interested to hear any further thoughts from you both on what to do in this scenario. Does this perhaps warrant a separate GitHub issue to ensure this is addressed at a later time?

enejb commented on packages/components/src/mobile/alt-text-settings/styles.scss:
```diff
@@ -0,0 +1,32 @@
+}
+
+.horizontalBorderDark {
+	border-color: $gray-70;
```
Do we not need not need 
```
border-top-width: $border-width;
border-bottom-width: $border-width;
```
here?

SiobhyB replied:
👋  @enejb, thank you for your comments and looking over the code here. :D 

I'm using [the usePreferredColorSchemeStyle hook](https://github.com/WordPress/gutenberg/blob/trunk/packages/compose/src/hooks/use-preferred-color-scheme-style/index.native.js) to change the styles depending on the user's colour scheme. It's my understanding that, with this hook, styles created under the light mode selector will always be applied and then we can override the ones we need to using the dark mode selector. So, the border widths here should be inherited from the light mode selector by default and that seems to be working in testing. 

Let me know if that seems right to you or if you still think more is needed here!

enejb replied:
Thanks for the explanation. That make sense :)

SiobhyB committed:
Convert "AltTextSettings" component into more generic "BottomSheetTextControl" component

This commit, converts the "AltTextSettings" component into more generic "BottomSheetTextControl" component by removing references to "alt" from main component and passing down props.

SiobhyB committed:
Created README.md for BottomSheetTextControl component

This commit introduces a README.md for the new BottomSheetTextControl component.

SiobhyB committed:
Formatting tweaks for README.md

SiobhyB committed:
Formatting tweaks for README.md

Fixes some issues with spacing.

SiobhyB commented:
> If we renamed it to something like BottomSheetTextControll and we could pass the different content that is specific to
current AltTextSettings into it would make it easier down the road the implement a similar feature on different blocks in the code base.

@enejb, this makes a lot of sense, thank you for the note! I've gone ahead to convert this to a more generic `BottomSheetTextControl` control now.

> On a related note it would be great to add a quick Readme to the new AltTextSettings component this helps other developers know how to use it :)

I've also gone ahead to create a README.md now. :)

@enejb, @guarani, when you have a chance, I'd be really grateful for your eyes on these changes. Thank you again for all the help and input you've already provided. 🙇‍♀️

SiobhyB committed:
retrigger checks

enejb commented:
Thanks for changes @SiobhyB and adding the documentation! ❤️ 

I wonder the current `FooterNote` part of `BottomSheetTextControl` is a bit too inflexible. 

Instead of passing in 3 different variables. We could pass in a single `FooterNote` component. That gets passed down into `BottomSheetTextControl`? 

This way in case we want to add 2 links for example or it would work out nicer. 

Also if we do take this approach a bit further we might also want to create a very simple `FooterNote` and `FooterNoteLink` components. 

(Maybe this is taking things too far) But in the end we would end up with something like. 

```
<BottomSheetTextControl
...
footerNote={
<FooterNote>
<>
    { __( 'Describe the purpose of the image. Leave empty if the image is purely decorative. ' ) }
    <FooterNoteLink url="'https://www.w3.org/WAI/tutorials/images/decision-tree/'">
      { __( 'What is alt text?' )}
    </FooterNoteLink>
    </>
</FooterNote>
}
/>
```

This way if we add footer notes to other bottom sheet control components they would all be the same styles. 
I think the down side is that we are introducing more components that we might not need.

What do you think?

SiobhyB committed:
Simplify BottomSheetTextControl

This commit seeks to simplify the BottomSheetTextControl component by splitting out two more, smaller components from it: 1. smallFooterMessageControl 2. smallFooterMessageControlLink

SiobhyB committed:
Merge branch 'trunk' into update/text-entry-for-alt-text-field

SiobhyB committed:
retrigger checks

SiobhyB committed:
Correct merge conflict

SiobhyB commented:
Thank you again @enejb, this makes total sense and is a really valuable lesson for me to be mindful about "future-proofing" changes. :) I've split out two smaller components to make this more re-usable: `SmallFooterMessageControl` and `SmallFooterMessageControlLink`. The reason for the naming here is that we already have [a FooterMessageControl control component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/footer-message-control/index.native.js) which uses different styling (with larger text) to what we're using here.

Let me know if this looks okay to you!

enejb commented:
Thanks for working on this. I have some more suggestions here :)

1. Instead of creating a new component `SmallFooterMessageControl` we can pass in a prop size="small" this aligns to the patterns that the new g2components are taking. See https://github.com/ItsJonQ/g2/tree/main/packages/components/src/Button#size. 
This would help with increasing the consistency across the different FooterMessage components. Right now we are introducing a new component that has a different size, colour, and text alignment. 

cc @iamthomasbishop Can we align the two footer message components to always be the same? The previous one is currently only being used in the columns component as well as the columns picker component. 

See (currently it looks like this)
<img width="395" alt="Screen Shot 2021-03-23 at 9 52 09 AM" src="https://user-images.githubusercontent.com/115071/112185565-959b6100-8bbd-11eb-96d8-f24d9cf78fa5.png">
<img width="401" alt="Screen Shot 2021-03-23 at 9 51 58 AM" src="https://user-images.githubusercontent.com/115071/112185570-96cc8e00-8bbd-11eb-86bb-02c6a9c3a1e5.png">

If we applied the same new footer style (new colour, fontsize and alignment) 
<img width="405" alt="Screen Shot 2021-03-23 at 9 54 26 AM" src="https://user-images.githubusercontent.com/115071/112185878-dc895680-8bbd-11eb-9950-d06b7cc7a667.png">
<img width="415" alt="Screen Shot 2021-03-23 at 9 54 15 AM" src="https://user-images.githubusercontent.com/115071/112185882-dd21ed00-8bbd-11eb-9b4e-71039ebb3072.png">

If we want to stick with the 2 different variations for now let's add the new one by passing in the correct attributes/props to the previous `FooterMessageControl` component instead of creating a new one. 
```
<FooterMessageControl size="small" label={ footerNote } textAlign="left" />
```
This helps developers choose the right component in the future since there will be only one :) 

2. I think we could omit the "control" suffix from `SmallFooterMessageLinkControl` name. Since the component doesn't control anything and it makes the component name more verbose. The same goes for the "Small" prefix. This way the new link component would be `FooterMessageLink` 

3. I noticed that the new footer has cut off the bottom parts of some of the letters. See
<img width="398" alt="Screen Shot 2021-03-23 at 9 58 45 AM" src="https://user-images.githubusercontent.com/115071/112186431-5d485280-8bbe-11eb-8585-147a927ec20b.png"> Notice `y` and `g `letters in the screenshot above.
I am not sure what changes are now causing this.

4. I think it would help if the link text was underlined to make it more appear more clickable. cc @iamthomasbishop 
See 
<img width="406" alt="Screen Shot 2021-03-23 at 10 02 29 AM" src="https://user-images.githubusercontent.com/115071/112186984-e3649900-8bbe-11eb-95f0-4699d2d06697.png">

5. 
```
<PanelBody style={ styles.textFooternote }>
<SmallFooterMessageControl
	value={ footerNote }
></SmallFooterMessageControl>
</PanelBody>
```

could be rewritten to be 
```
{ footerNote && <PanelBody style={ styles.textFooternote }>
  <SmallFooterMessageControl
	  value={ footerNote }
  ></SmallFooterMessageControl>
</PanelBody> }
```

This way the `FooterMessageControl` would only show up if there is a footer note argument. 
		
6. It would be great to also add a readme file for the new FooterMessageLink component. :)

SiobhyB committed:
Update "SmallFooterMessageLinkControl" to "FooterMessageLink"

This commit updates the name of "SmallFooterMessageLinkControl" to "FooterMessageLink" in order to be clearer and more readable.

SiobhyB committed:
Update "SmallFooterMessageLinkControl" to "FooterMessageLink"

This commit updates the name of "SmallFooterMessageLinkControl" to "FooterMessageLink" in order to be clearer and more readable.

SiobhyB committed:
Delete small-footer-message-control-link.native.js

SiobhyB committed:
Remove small-footer-message-control

SiobhyB committed:
Update Footer Message Control

SiobhyB commented:
Thanks again for your help! I've implemented the following from your list:

`2.` Done, I've changed `SmallFooterMessageLinkControl` to `FooterMessageLink`. ✅
`5.` Done, thank you for the tip! :D ✅

And am either working on or have some follow up questions on the following:

`1.` I converted `SmallFooterMessageControl` to `FooterMessageControl` and passed the `text-align` prompt, as you suggested. Adding `size="small"` doesn't work to change the font size, however. I'll wait to hear from @iamthomasbishop on his thoughts around aligning the styles before digging further into this.
`3.` I'm still working to replicate this.
`4.` I'll await @iamthomasbishop's feedback on the design change you suggested, also.
`6.` I placed the `FooterMessageLink` component under the `bottom-sheet` folder as I see that's where a lot of other smaller components for mobile live. There aren't any other READMEs for these smaller components, so am not sure where to place a new one exactly. Should I perhaps create a separate GitHub issue regarding these components with missing READMEs?

SiobhyB committed:
Add README to FooterMessageLink component

This commit moves FooterMessageLink to a separate folder for the purpose of adding a README

SiobhyB committed:
Revert accidental commit

This reverts an accidental commit, where I'd made changes in "link-settings/index.native.js" as part of general testing. I also remove "textAlign="left"" from a FooterMessageControl component in this file, as this is now the default.

SiobhyB commented:
Following a chat with @iamthomasbishop, I've gone ahead to complete the following:

* I've changed the style of `FooterMessageControl` so that its text is aligned to the left and the size is now slightly smaller, at 12px. The styles of the footer message in the Columns block and the Image block's settings now match. :) This should also resolve the issue with padding.
* We're going to keep the style of the link as-is for now, with the option of opening a new PR if we'd like to change this at a later stage.
* I've added a README for the `FooterMessageLink` component.

**Android:**

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/112872046-d3b1ec80-90b7-11eb-9c4a-7aea3a5a85ac.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/112871867-9c434000-90b7-11eb-9f13-006a14bfe919.png" width="100%"> |

**iOS:**

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/112876262-348ff380-90bd-11eb-8637-f0878ed94dea.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/112876159-0e6a5380-90bd-11eb-9b87-df679a14cb4d.png" width="100%">

New issues created following the discussions here:

* Low Contrast Between Text Colour and Background in Footer: #29866. 
* Missing READMEs for Some Mobile Components: #30355.

@enejb, could I bother you for another review, please? Thank you in advance. 🙇‍♀️

enejb commented:
Nicely done @SiobhyB! 

I tested this on the android and ios simulators and it worked well for me. (lite and dark modes) 
Thank you for all the different code changes. I think this is ready to ship :shipit: !

iamthomasbishop reviewed (changes requested):
@SiobhyB This is looking good, just one tiny thing I noticed in your last round of screenshots. It looks like the text label might be getting clipped by the container boundary:

![Screen Shot 2021-04-01 at 10 29 50 AM](https://user-images.githubusercontent.com/1200432/113318424-1224f280-92d6-11eb-9df5-cb0a49fa8cff.png)

SiobhyB commented:
@iamthomasbishop, thanks for your review! I spent a bit of time looking into this and it appears to be Android-specific, with the bottom parts of some text being visible at the top when multiple lines are added. I couldn't find a quick fix, as testing will be required across different size screens in both iOS and Android, but have created a new GitHub issue for this: https://github.com/WordPress/gutenberg/issues/30457

Do you think it'd be okay to merge this PR and then I'll work on that issue separately when I'm back from AFK?

iamthomasbishop commented:
> Do you think it'd be okay to merge this PR and then I'll work on that issue separately when I'm back from AFK?

@SiobhyB That's fine, thank you!

hypest commented:
~Just to be sure, since there are bridge interface changes, we have to merge the WPAndroid side at the same time, right? If so, we'll need to get [the WPAndroid PR](https://github.com/wordpress-mobile/WordPress-Android/pull/14052/) CI green and approved before merging 👍~

Oops, I mixed up two different efforts, there are no bridge changes in this one so, will go ahead and merge (and update+merge the gutenberg-mobile PR too). Thanks for the work here @SiobhyB!
